espnet2/enh/separator/tcn_separator.py

```diff
@@ -114,6 +114,79 @@ def forward(
 
         return masked, ilens, others
 
+    def forward_streaming(self, input_frame: torch.Tensor, buffer=None):
+        # input_frame: B, 1, N
+
+        B, _, N = input_frame.shape
+
+        receptive_field = self.tcn.receptive_field
+
+        if buffer is None:
+            buffer = torch.zeros((B, receptive_field, N), device=input_frame.device)
+
+        buffer = torch.roll(buffer, shifts=-1, dims=1)
+        buffer[:, -1, :] = input_frame[:, 0, :]
+
+        masked, ilens, others = self.forward(buffer, None)
+
+        masked = [m[:, -1, :].unsqueeze(1) for m in masked]
+
+        return masked, buffer, others
+
     @property
     def num_spk(self):
         return self._num_spk
+
+
+if __name__ == "__main__":
+    import time
+
+    import humanfriendly
+
+    SEQ_LEN = 10000
+    num_spk = 2
+
+    separator = TCNSeparator(
+        input_dim=128,
+        num_spk=2,
+        layer=6,
+        stack=3,
+        bottleneck_dim=32,
+        hidden_dim=64,
+        kernel=3,
+        causal=True,
+        norm_type="cLN",
+    )
+
+    separator.eval()
+
+    print(
+        f"Number of parameters: "
+        f"{humanfriendly.format_size(sum(p.numel() for p in separator.parameters()))}"
+    )
+
+    input_feature = torch.randn((3, SEQ_LEN, 128))
+    ilens = torch.LongTensor([SEQ_LEN, SEQ_LEN, SEQ_LEN])
+
+    with torch.no_grad():
+        start = time.time()
+        seq_output, _, _ = separator.forward(input_feature, ilens=ilens)
+        end = time.time()
+        print(f"seqeunce processing, time cost: {end - start}")
+
+        start = time.time()
+        state = None
+        stream_outputs = []
+        for i in range(SEQ_LEN):
+            frame = input_feature[:, i : i + 1, :]
+            frame_out, state, _ = separator.forward_streaming(frame, state)
+            stream_outputs.append(frame_out)
+        end = time.time()
+        print(f"streaming processing, time cost: {end - start}")
+        for i in range(SEQ_LEN):
+            for s in range(num_spk):
+                torch.testing.assert_allclose(
+                    stream_outputs[i][s], seq_output[s][:, i : i + 1, :]
+                )
+
+    print("Streaming OKey")
```

Comment:
We can remove this part as the unit test already covers it.

Reply:
Removed.